### ISSUE TITLE
vm templating corrupt guest image?

### ISSUE THREAD

**Issue by wenbinzeng** [kata-containers/qemu#5]
I am testing the migration capability 'bypass-shared-memory' introduced in 8ba3a0fd7209ee16da1d82ed966d9810747cb7fa, found that the guest images always got corrupted on the cloned the guests. I wonder if I did things wrong?

Here is what I do:
1. I built qemu branch qemu-lite-2.11.0

2. start a vm using comman line:
--------------------------------------
qemu-system-x86_64 -machine q35,accel=kvm,kernel_irqchip,smm=off,nosmbus,nosata,nopit,static-prt,nofw -cpu host -m 1024,maxmem=20G,slots=2 -smp 1 -name guest-20115 -device pci-bridge,id=pci-bridge1,chassis_nr=1 -kernel /local/guest-kernels/vmlinux -append root=/dev/vda rw rootfstype=ext4 tsc=reliable no_timer_check rcupdate.rcu_expedited=1 pci=lastbus=0 reboot=t panic=-1  -no-user-config -nodefaults -no-hpet -nographic -object memory-backend-file,id=mem,size=1024M,mem-path=/dev/shm/qemu-lite/guest_mem,share=on -drive file=/data/centos7-base.qcow2,format=qcow2,if=virtio,aio=threads -netdev tap,id=hostnet0,vhost=on,script=no,downscript=no,ifname=veth58 -device virtio-net-pci,netdev=hostnet0,mac=52:54:00:12:34:99 -device virtio-serial-pci,id=virtio-serial0,disable-modern -monitor unix:/var/qemu-lite/monitor.guest-20115,server,nowait -chardev stdio,id=charconsole0 -device virtconsole,chardev=charconsole0,id=console0
-----------------------------------------

3. in qemu monitor, make a vm template as follows:
---------------------------------------------------------
(qemu) migrate_set_capability bypass-shared-memory on
(qemu) stop
(qemu) migrate "exec:cat > /dev/shm/qemu-lite/guest_template"
(qemu) quit
-------------------------------------------------------

4. clone a vm from the template, changed two lines of the above qemu command line like below:
----------------------------------------------------
-object memory-backend-file,id=mem,size=1024,mem-path=/dev/shm/qemu-lite/guest_mem,share=off -incoming "exec:cat /dev/shm/qemu-lite/guest_template"
-----------------------------------------------------

5. resume the clone vm in qemu monitor i.e.
(qemu) c

6. in the guest os, I modifed a file , say, /home/abc.txt.
Then shutdown the guest, and start it again by repeating step 4&5. 
When I check the file /home/abc.txt again, guest FS seems corrupted:
-------------------------------------------
# cat /home/abc.txt
cat: /home/abc.txt: Structure needs cleaning
------------------------------------------

This is constantly reproducible. Am I doing anything wrong?

Please advice, thanks in advance!

**Comment by bergwolf** [comment]
hi, vm template shares the initial guest memory. Every time you create a new guest, the very initial guest memory is the same, -- IWO, the qcow2 rootfs in memory metadata is the same. So if you modify a file in step 6, next time you create a new guest, the guest kernel still holds the unchanged metadata for the file. And the ondisk data is changed (thus does not match the in-memory metadata), so kernel can either report a data corruption, or even crash.

**Comment by wenbinzeng** [comment]
@bergwolf 
Thank you very much for your clarification. One more question:
In the above step 3, "bypass-shared-memory" need to be turned on undoubtedly. My question is, do I need to turn it on too in step 4 i.e. before executing "qemu-system-x86_64 ... -incoming ..."?

**Comment by bergwolf** [comment]
@wenbinzeng No, you don't need it in step 4. `bypass-shared-memory` is a migration capability that tells qemu to bypass shared memory when doing migration save. In step 4, qemu is actually doing migration load.

**Comment by wenbinzeng** [comment]
@bergwolf 
Very much appreciated for you explanation!

Both step 2 and step 4 have set below memory-backend-file:
mem-path=/dev/shm/qemu-lite/guest_mem

What is in this file?
Can I run multiple qemu guests on this particular file?
Can I delete this file after step 3 and before step 4?

**Comment by bergwolf** [comment]
> What is in this file?

It is the guest memory file we want to share among all guests created from vm template. 

> Can I run multiple qemu guests on this particular file?

Yes.

> Can I delete this file after step 3 and before step 4?

No, you can't. The file is needed for all guests to share the template vm memory.

**Comment by wenbinzeng** [comment]
In vm templating scenario, what is the proper size of the shared memory-backend-file? Should it be exactly same to the virtual machine's memory size? For example, the following template vm has 1GB memory, must the memory-backend-file object size be 1GB too?

-m 1024,maxmem=20G,slots=2 
-object memory-backend-file,id=mem,size=1024M,mem-path=/dev/shm/qemu-lite/guest_mem,share=on
